File: Source/Libraries/openHistorianModule/remoteBinaryStream.py
#******************************************************************************************************
#  remoteBinaryStream.py - Gbtc
#
#  Copyright © 2021, Grid Protection Alliance.  All Rights Reserved.
#
#  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
#  the NOTICE file distributed with this work for additional information regarding copyright ownership.
#  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may not use this
#  file except in compliance with the License. You may obtain a copy of the License at:
#
#      http://opensource.org/licenses/MIT
#
#  Unless agreed to in writing, the subject software distributed under the License is distributed on an
#  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
#  License for the specific language governing permissions and limitations.
#
#  Code Modification History:
#  ----------------------------------------------------------------------------------------------------
#  01/31/2021 - J. Ritchie Carroll
#       Generated original version of source code.
#
#******************************************************************************************************

from encoding7Bit import encoding7Bit
from streamEncoding import streamEncoding
from common import ByteSize, Validate
from uuid import UUID
import numpy as np

class remoteBinaryStream:
    """
    Establishes buffered I/O around a base stream, e.g., a socket
    """

    BufferSize = 1420

    def __init__(self, stream: streamEncoding):
        self.stream = stream
        self.receiveBuffer = bytearray(remoteBinaryStream.BufferSize)
        self.sendBuffer = bytearray(remoteBinaryStream.BufferSize)
        self.sendLength = 0
        self.receiveLength = 0
        self.receivePosition = 0

    @property
    def SendBufferFreeSpace(self) -> int:
        return remoteBinaryStream.BufferSize - self.sendLength

    @property
    def ReceiveBufferAvailable(self) -> int:
        return self.receiveLength - self.receivePosition

    def Flush(self):
        if self.sendLength <= 0:
            return

        self.stream.Write(self.sendBuffer, 0, self.sendLength)
        self.sendLength = 0

    def Read(self, buffer: bytearray, offset: int, count: int) -> int:
        if count <= 0:
            return 0;

        receiveBufferLength = self.ReceiveBufferAvailable

        # Check if there is enough in the receive buffer to handle request
        if count <= receiveBufferLength:
            for i in range(count):
                buffer[offset + i] = self.receiveBuffer[self.receivePosition + i]

            self.receivePosition += count
            return count

        originalCount = count

        # Empty existing receive buffer
        if receiveBufferLength > 0:
            for i in range(receiveBufferLength):
                buffer[offset + i] = self.receiveBuffer[self.receivePosition + i]

            self.receivePosition = 0
            self.receiveLength = 0
            offset += receiveBufferLength
            count -= receiveBufferLength
        
        # If more than 100 bytes remain, skip receive buffer 
        # and copy directly to the destination
        if count > 100:
            # Loop since socket reads can return partial results
            while count > 0:
                receiveBufferLength = self.stream.Read(buffer, offset, count)

                if receiveBufferLength == 0:
                    raise RuntimeError("End of stream")

                offset += receiveBufferLength
                count -= receiveBufferLength            
        else:
            # With fewer than 100 bytes requested, fill receive buffer 
            # then copy to destination
            prebufferLength = remoteBinaryStream.BufferSize
            self.receiveLength = 0

            while self.receiveLength < count:
                receiveBufferLength = self.stream.Read(self.receiveBuffer, self.receiveLength, prebufferLength)
                
                if receiveBufferLength == 0:
                    raise RuntimeError("End of stream")

                self.receiveLength += receiveBufferLength
                prebufferLength -= receiveBufferLength

            for i in range(count):
                buffer[offset + i] = self.receiveBuffer[i]

            self.receivePosition = count

        return originalCount

    def Write(self, buffer: bytearray, offset: int, count: int) -> int:
        if self.SendBufferFreeSpace < count:
            self.Flush()

        if count > 100:
            self.Flush()
            self.stream.Write(buffer, offset, count)
        else:
            for i in range(count):
                self.sendBuffer[self.sendLength + i] = buffer[offset + i]
            
            self.sendLength += count

        return count

    def ReadAll(self, buffer: bytearray, position: int, length: int):
        """
        Reads all of the provided bytes. Will not return prematurely, 
        continues to execute a `Read` command until the entire
        `length` has been read.
        """
        Validate.Parameters(buffer, position, length)

        while length > 0:
            bytesRead = self.Read(buffer, position, length)
            
            if bytesRead == 0:
                raise RuntimeError("End of stream")

            length -= bytesRead
            position += bytesRead

    def ReadBytes(self, count: int) -> bytes:
        buffer = bytearray(count)
        self.ReadAll(buffer, 0, count)
        return bytes(buffer)

    def ReadNextBuffer(self) -> bytes:
        return self.ReadBytes(self.Read7BitUInt32())

    def ReadString(self) -> str:
        return self.ReadNextBuffer().decode("utf8")
    
    def ReadGuid(self) -> UUID:
        return UUID(bytes_le=self.ReadBytes(16))

    def Read7BitInt32(self) -> np.int32:
        return encoding7Bit.ReadInt32(self.ReadByte)
        
    def Read7BitUInt32(self) -> np.uint32:
        return encoding7Bit.ReadUInt32(self.ReadByte)

    def Read7BitInt64(self) -> np.int64:
        return encoding7Bit.ReadInt64(self.ReadByte)

    def Read7BitUInt64(self) -> np.uint64:
        return encoding7Bit.ReadUInt64(self.ReadByte)

    def ReadByte(self) -> np.uint8:
        size = ByteSize.UINT8

        if self.receivePosition < self.receiveLength:
            value = self.receiveBuffer[self.receivePosition]
            self.receivePosition += size
            return np.uint8(value)

        return self.stream.ReadByte()

    def WriteByte(self, value: np.uint8) -> int:
        size = ByteSize.UINT8

        if self.sendLength < remoteBinaryStream.BufferSize:
            self.sendBuffer[self.sendLength] = value
            self.sendLength += size
            return size
        
        return self.stream.WriteByte(value)

    def ReadBoolean(self) -> bool:
        return self.ReadByte() != 0

    def WriteBoolean(self, value: bool) -> int:
        if value:
            self.WriteByte(1)
        else:
            self.WriteByte(0)

        return 1

    def ReadInt16(self) -> np.int16:
        size = ByteSize.INT16

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.int16)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadInt16()

    def WriteInt16(self, value: np.int16) -> int:
        size = ByteSize.INT16

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little", signed=True)

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteInt16(value)

    def ReadUInt16(self) -> np.uint16:
        size = ByteSize.UINT16

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.uint16)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadUInt16()

    def WriteUInt16(self, value: np.uint16) -> int:
        size = ByteSize.UINT16

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little")

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteUInt16(value)

    def ReadInt32(self) -> np.int32:
        size = ByteSize.INT32

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.int32)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadInt32()

    def WriteInt32(self, value: np.int32) -> int:
        size = ByteSize.INT32

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little", signed=True)

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteInt32(value)

    def ReadUInt32(self) -> np.uint32:
        size = ByteSize.UINT32

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.uint32)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadUInt32()

    def WriteUInt32(self, value: np.uint32) -> int:
        size = ByteSize.UINT32

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little")

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteUInt32(value)

    def ReadInt64(self) -> np.int64:
        size = ByteSize.INT64

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.int64)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadInt64()

    def WriteInt64(self, value: np.int64) -> int:
        size = ByteSize.INT64

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little", signed=True)

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteInt64(value)

    def ReadUInt64(self) -> np.uint64:
        size = ByteSize.UINT64

        if self.receivePosition <= self.receiveLength - size:
            value = np.frombuffer(self.receiveBuffer[self.receivePosition:self.receivePosition + size], np.uint64)[0]
            self.receivePosition += size
            return value

        return self.stream.ReadUInt64()

    def WriteUInt64(self, value: np.uint64) -> int:
        size = ByteSize.UINT64

        if self.sendLength <= remoteBinaryStream.BufferSize - size:
            buffer = value.to_bytes(size, "little")

            for i in range(size):
                self.sendBuffer[self.sendLength + i] = buffer[i]

            self.sendLength += size
            return size

        return self.stream.WriteUInt64(value)

    def __sendBufferRead(self, length: int) -> bytes:
        buffer = bytearray(length)

        for i in range(length):
            buffer[i] = self.receiveBuffer[self.receivePosition + i]

        self.receivePosition += length
        return bytes(buffer)

    def __sendBufferWrite(self, buffer: bytes) -> int:
        length = len(buffer)

        for i in range(length):
            self.sendBuffer[self.sendLength + i] = buffer[i]

        self.sendLength += length
        return length

    def Write7BitUInt64(self, value: np.uint64) -> int:
        if self.sendLength <= remoteBinaryStream.BufferSize - ByteSize.ENC7BIT:
            stream = streamEncoding(self.__sendBufferRead, self.__sendBufferWrite)
            return stream.Write7BitUInt64(value)

        return self.stream.Write7BitUInt64(value)

    def Read7BitUInt64(self) -> np.uint64:
        if self.receivePosition <= self.receiveLength - ByteSize.ENC7BIT:
            stream = streamEncoding(self.__sendBufferRead, self.__sendBufferWrite)
            return stream.Read7BitUInt64()

        return self.stream.Read7BitUInt64()

File: Source/Libraries/GSF.SortedTreeStore/IO/RemoteBinaryStream.cs
//******************************************************************************************************
//  RemoteBinaryStream.cs - Gbtc
//
//  Copyright © 2014, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  12/08/2012 - Steven E. Chisholm
//       Generated original version of source code. 
//       
//
//******************************************************************************************************

using System;
using System.IO;
using GSF.IO;
using GSF.Threading;

namespace GSF.Net
{
    /// <summary>
    /// A binary stream that works on some kind of piped stream.
    /// This means the reader of one stream 
    /// is connected to the writer of another stream.
    /// </summary>
    public class RemoteBinaryStream
        : BinaryStreamBase
    {
        private const int BufferSize = 1420;
        private int m_receivePosition;
        private int m_receiveLength;
        private int m_sendLength;
        private readonly byte[] m_receiveBuffer;
        private readonly byte[] m_sendBuffer;

        private readonly Stream m_stream;
        private readonly WorkerThreadSynchronization m_workerThreadSynchronization;

        /// <summary>
        /// Creates a <see cref="RemoteBinaryStream"/>
        /// </summary>
        /// <param name="stream">the underlying stream to wrap</param>
        /// <param name="workerThreadSynchronization">the synchronization object</param>
        public RemoteBinaryStream(Stream stream, WorkerThreadSynchronization workerThreadSynchronization = null)
        {
            if (!BitConverter.IsLittleEndian)
                throw new Exception("BigEndian processors are not supported");

            if (workerThreadSynchronization is null)
                workerThreadSynchronization = new WorkerThreadSynchronization();

            m_workerThreadSynchronization = workerThreadSynchronization;
            m_receiveBuffer = new byte[BufferSize];
            m_sendBuffer = new byte[BufferSize];
            m_sendLength = 0;
            m_receiveLength = 0;
            m_receivePosition = 0;
            m_stream = stream;
        }

        /// <summary>
        /// Gets the <see cref="WorkerThreadSynchronization"/>. 
        /// This context will be entered when communcating to the socket layer.
        /// </summary>
        public WorkerThreadSynchronization WorkerThreadSynchronization => m_workerThreadSynchronization;

        private int SendBufferFreeSpace => BufferSize - m_sendLength;

        protected int ReceiveBufferAvailable => m_receiveLength - m_receivePosition;

        public override bool CanWrite => true;

        public override long Length => throw new NotSupportedException();

        public override long Position
        {
            get => throw new NotSupportedException();
            set => throw new NotSupportedException();
        }

        public override void Flush()
        {
            if (m_sendLength <= 0)
                return;

            m_workerThreadSynchronization.BeginSafeToCallbackRegion();
            try
            {
                m_stream.Write(m_sendBuffer, 0, m_sendLength);
                m_stream.Flush();
            }
            finally
            {
                m_workerThreadSynchronization.EndSafeToCallbackRegion();
            }
            m_sendLength = 0;
        }

        public override void SetLength(long value)
        {
            throw new NotSupportedException();
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            if (count <= 0)
                return 0;

            int receiveBufferLength = ReceiveBufferAvailable;

            //if there is enough in the receive buffer to fulfill this request
            if (count <= receiveBufferLength)
            {
                Array.Copy(m_receiveBuffer, m_receivePosition, buffer, offset, count);
                m_receivePosition += count;
                return count;
            }
            int originalCount = count;

            //first empty the receive buffer.
            if (receiveBufferLength > 0)
            {
                Array.Copy(m_receiveBuffer, m_receivePosition, buffer, offset, receiveBufferLength);
                m_receivePosition = 0;
                m_receiveLength = 0;
                offset += receiveBufferLength;
                count -= receiveBufferLength;
            }

            //if still asking for more than 100 bytes, skip the receive buffer 
            //and copy directly to the destination
            if (count > 100)
            {
                //Loop since ReceiveFromSocket can return parial results.
                while (count > 0)
                {
                    m_workerThreadSynchronization.BeginSafeToCallbackRegion();
                    try
                    {
                        receiveBufferLength = m_stream.Read(buffer, offset, count);
                    }
                    finally
                    {
                        m_workerThreadSynchronization.EndSafeToCallbackRegion();
                    }


                    if (receiveBufferLength == 0)
                        throw new EndOfStreamException();
                    offset += receiveBufferLength;
                    count -= receiveBufferLength;
                }
                return originalCount;
            }
            else
            {
                //With fewer than 100 bytes requested, 
                //first fill up the receive buffer, 
                //then copy this to the destination.
                int prebufferLength = m_receiveBuffer.Length;
                m_receiveLength = 0;
                while (m_receiveLength < count)
                {
                    m_workerThreadSynchronization.BeginSafeToCallbackRegion();
                    try
                    {
                        receiveBufferLength = m_stream.Read(m_receiveBuffer, m_receiveLength, prebufferLength);
                    }
                    finally
                    {
                        m_workerThreadSynchronization.EndSafeToCallbackRegion();
                    }
                    if (receiveBufferLength == 0)
                        throw new EndOfStreamException();
                    m_receiveLength += receiveBufferLength;
                    prebufferLength -= receiveBufferLength;
                }
                Array.Copy(m_receiveBuffer, 0, buffer, offset, count);
                m_receivePosition = count;
                return originalCount;
            }
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                WorkerThreadSynchronization.BeginSafeToCallbackRegion();
            }
            base.Dispose(disposing);
        }

        public override void Write(byte value)
        {
            if (m_sendLength < BufferSize)
            {
                m_sendBuffer[m_sendLength] = value;
                m_sendLength++;
                return;
            }
            base.Write(value);
        }

        public unsafe override void Write(long value)
        {
            if (m_sendLength <= BufferSize - 8)
            {
                fixed (byte* lp = m_sendBuffer)
                {
                    *(long*)(lp + m_sendLength) = value;
                    m_sendLength += 8;
                    return;
                }
            }
            base.Write(value);
        }

        public unsafe override void Write(int value)
        {
            if (m_sendLength <= BufferSize - 4)
            {
                fixed (byte* lp = m_sendBuffer)
                {
                    *(int*)(lp + m_sendLength) = value;
                    m_sendLength += 4;
                    return;
                }
            }
            base.Write(value);
        }

        public unsafe override void Write7Bit(ulong value)
        {
            const int size = 9;
            if (m_sendLength <= BufferSize - 9)
            {
                fixed (byte* lp = m_sendBuffer)
                {
                    byte* stream = lp + m_sendLength;
                    if (value < 128)
                    {
                        stream[0] = (byte)value;
                        m_sendLength += 1;
                        return;
                    }
                    stream[0] = (byte)(value | 128);
                    if (value < 128 * 128)
                    {
                        stream[1] = (byte)(value >> 7);
                        m_sendLength += 2;
                        return;
                    }
                    stream[1] = (byte)((value >> 7) | 128);
                    if (value < 128 * 128 * 128)
                    {
                        stream[2] = (byte)(value >> 14);
                        m_sendLength += 3;
                        return;
                    }
                    stream[2] = (byte)((value >> 14) | 128);
                    if (value < 128 * 128 * 128 * 128)
                    {
                        stream[3] = (byte)(value >> 21);
                        m_sendLength += 4;
                        return;
                    }
                    stream[3] = (byte)((value >> (7 + 7 + 7)) | 128);
                    if (value < 128L * 128 * 128 * 128 * 128)
                    {
                        stream[4] = (byte)(value >> (7 + 7 + 7 + 7));
                        m_sendLength += 5;
                        return;
                    }
                    stream[4] = (byte)((value >> (7 + 7 + 7 + 7)) | 128);
                    if (value < 128L * 128 * 128 * 128 * 128 * 128)
                    {
                        stream[5] = (byte)(value >> (7 + 7 + 7 + 7 + 7));
                        m_sendLength += 6;
                        return;
                    }
                    stream[5] = (byte)((value >> (7 + 7 + 7 + 7 + 7)) | 128);
                    if (value < 128L * 128 * 128 * 128 * 128 * 128 * 128)
                    {
                        stream[6] = (byte)(value >> (7 + 7 + 7 + 7 + 7 + 7));
                        m_sendLength += 7;
                        return;
                    }
                    stream[6] = (byte)((value >> (7 + 7 + 7 + 7 + 7 + 7)) | 128);
                    if (value < 128L * 128 * 128 * 128 * 128 * 128 * 128 * 128)
                    {
                        stream[7] = (byte)(value >> (7 + 7 + 7 + 7 + 7 + 7 + 7));
                        m_sendLength += 8;
                        return;
                    }
                    stream[7] = (byte)(value >> (7 + 7 + 7 + 7 + 7 + 7 + 7) | 128);
                    stream[8] = (byte)(value >> (7 + 7 + 7 + 7 + 7 + 7 + 7 + 7));
                    m_sendLength += 9;
                    return;
                }

            }
            base.Write7Bit(value);
        }

        public override byte ReadUInt8()
        {
            if (m_receivePosition < m_receiveLength)
            {
                byte value = m_receiveBuffer[m_receivePosition];
                m_receivePosition++;
                return value;
            }
            return base.ReadUInt8();
        }

        public unsafe override int ReadInt32()
        {
            if (m_receivePosition <= m_receiveLength - 4)
            {
                fixed (byte* lp = m_receiveBuffer)
                {
                    int value = *(int*)(lp + m_receivePosition);
                    m_receivePosition += 4;
                    return value;
                }
            }
            return base.ReadInt32();
        }

        public unsafe override long ReadInt64()
        {
            if (m_receivePosition <= m_receiveLength - 8)
            {
                fixed (byte* lp = m_receiveBuffer)
                {
                    long value = *(long*)(lp + m_receivePosition);
                    m_receivePosition += 8;
                    return value;
                }
            }
            return base.ReadInt64();
        }

        public unsafe override ulong Read7BitUInt64()
        {
            if (m_receivePosition <= m_receiveLength - 9)
            {
                fixed (byte* lp = m_receiveBuffer)
                {
                    byte* stream = lp + m_receivePosition;
                    ulong value11;
                    value11 = stream[0];
                    if (value11 < 128)
                    {
                        m_receivePosition += 1;
                        return value11;
                    }
                    value11 ^= (ulong)stream[1] << 7;
                    if (value11 < 128 * 128)
                    {
                        m_receivePosition += 2;
                        return value11 ^ 0x80;
                    }
                    value11 ^= (ulong)stream[2] << (7 + 7);
                    if (value11 < 128 * 128 * 128)
                    {
                        m_receivePosition += 3;
                        return value11 ^ 0x4080;
                    }
                    value11 ^= (ulong)stream[3] << (7 + 7 + 7);
                    if (value11 < 128 * 128 * 128 * 128)
                    {
                        m_receivePosition += 4;
                        return value11 ^ 0x204080;
                    }
                    value11 ^= (ulong)stream[4] << (7 + 7 + 7 + 7);
                    if (value11 < 128L * 128 * 128 * 128 * 128)
                    {
                        m_receivePosition += 5;
                        return value11 ^ 0x10204080L;
                    }
                    value11 ^= (ulong)stream[5] << (7 + 7 + 7 + 7 + 7);
                    if (value11 < 128L * 128 * 128 * 128 * 128 * 128)
                    {
                        m_receivePosition += 6;
                        return value11 ^ 0x810204080L;
                    }
                    value11 ^= (ulong)stream[6] << (7 + 7 + 7 + 7 + 7 + 7);
                    if (value11 < 128L * 128 * 128 * 128 * 128 * 128 * 128)
                    {
                        m_receivePosition += 7;
                        return value11 ^ 0x40810204080L;
                    }
                    value11 ^= (ulong)stream[7] << (7 + 7 + 7 + 7 + 7 + 7 + 7);
                    if (value11 < 128L * 128 * 128 * 128 * 128 * 128 * 128 * 128)
                    {
                        m_receivePosition += 8;
                        return value11 ^ 0x2040810204080L;
                    }
                    value11 ^= (ulong)stream[8] << (7 + 7 + 7 + 7 + 7 + 7 + 7 + 7);
                    m_receivePosition += 9;
                    return value11 ^ 0x102040810204080L;
                }
            }
            return base.Read7BitUInt64();
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            if (SendBufferFreeSpace < count)
                Flush();

            if (count > 100)
            {
                Flush();
                m_workerThreadSynchronization.BeginSafeToCallbackRegion();
                try
                {
                    m_stream.Write(buffer, offset, count);
                }
                finally
                {
                    m_workerThreadSynchronization.EndSafeToCallbackRegion();
                }
            }
            else
            {
                Array.Copy(buffer, offset, m_sendBuffer, m_sendLength, count);
                m_sendLength += count;
            }
        }

        public override bool CanRead => true;

        public override bool CanSeek => false;
    }
}